### PR TITLE
Pipes - Irene D - Munchies

### PR DESCRIPTION
**Pull request by idevera**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? | I used Postman and HTTParty to query the API. I also read the Edamam documentation and used the example url to try and make an API call. I also requested an app key and id.  |
| Describe your API Wrapper.  How did you decide on the methods you created? | I had to create a method that would submit a query and return all the recipe data from the API and create Recipe instances. I also created another method to be able to find a particular recipe, return the information, and create one Recipe instance |
| Describe an edge case or failure case test you wrote for your API Wrapper. |  A failure case is that the APIWrapper will raise an error if the response to the API is anything but a 200 response. Another case was if a uri is bad, it will also raise an error. |
| Explain how VCR aids in testing an API. | VCR testing records the expected response of the API and saves it as a YML file. Any time the test is ran again, the test is ran from the file instead of calling the API |
| What is the Heroku URL of your deployed application? | https://munchies-recipes.herokuapp.com/  |

### CODE REVIEW

File: lib/edamam_api_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+class EdamamApiWrapper
+  BASE_URL = "https://api.edamam.com/search"
+  BASE_URI = "http://www.edamam.com/ontologies/edamam.owl%23recipe_"
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:3]: These two variables are confusingly similar. I would probably rename the second to something like `RECIPE_URI_PREFIX` to make things more visually distinct.

File: test/controllers/recipes_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        get recipes_path, params: input
+
+        must_respond_with :success
+      end
+    end
+
+    it "redirects back if search is blank" do
+      input = {
+        q: ""
+      }
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:28]: A few other tests I would be interested in:

1. What if the search returns no results (for example, a search for "dfaljdfalflaksjflaksj")?
1. Does the page still render if paging parameters are provided?
1. What if the paging parameters are invalid (negative number, etc)?
1. What if the paging parameters take you past the end of the available search results?

File: test/lib/edamam_api_wrapper_test.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        response.each do |recipe|
+          recipe.must_be_kind_of Recipe
+        end
+        response.length.must_be :>, 0
+      end
+    end
+
+    it "raises an ApiError when the app_id and app_key is bad" do
+      VCR.use_cassette("bad_tokens") do
+        proc {
```
Comment (droberts-sea) [inline test/lib/edamam_api_wrapper_test.rb:18]: I like this test! The one other case I would like to see is a search that returns no results.

File: test/lib/edamam_api_wrapper_test.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        all_recipes = EdamamApiWrapper.search("chicken")
+        recipe = all_recipes[0]
+        recipe_data = EdamamApiWrapper.find_recipe(recipe.uri_id)
+        recipe_data.must_be_kind_of Recipe
+      end
+    end
+
+    it "raises an ArgumentError with a bad uri" do
+      VCR.use_cassette("bad_uri") do
+        proc { EdamamApiWrapper.find_recipe("BadURI") }.must_raise ArgumentError
```
Comment (droberts-sea) [inline test/lib/edamam_api_wrapper_test.rb:38]: Good failure case

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | yes |
| Errors are reported to the user | yes |
| API Wrapper to handle the API requests | yes |
| Controller testing | yes - could be expanded somewhat |
| Lib testing | yes |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** |  |
| Link to deployed app on Heroku | yes |
|  **Overall** | Excellent job overall! This site is attractive and responsive, easy to navigate and for the most part well-tested. Keep up the hard work! |
